### ISSUE TITLE
Multiple conflicting errors in relation with using `brew cask ...`

### ISSUE THREAD

**Issue by moreaki** [caskroom/homebrew-cask#27378]
Through normal operations with homebrew, I ended up learning that casks supposedly have moved (https://github.com/caskroom/homebrew-cask/issues/21913); fine. I tried to understand and follow all the differing pieces of information that people gave in that thread and also read reports in this project's issue threads. To no avail!

My goal is to get a back a working brew/cask environment on my Mac, so I can continue using it the way it was designed. Currently, I get the following errors:

```
$ brew cask list
Error: No available formula with the name "brew-cask"
Warning: The default Caskroom location has moved to /usr/local/Caskroom.

Please migrate your Casks to the new location and delete /opt/homebrew-cask/Caskroom,
or if you would like to keep your Caskroom at /opt/homebrew-cask/Caskroom, add the
following to your HOMEBREW_CASK_OPTS:

  --caskroom=/opt/homebrew-cask/Caskroom

For more details on each of those options, see https://github.com/caskroom/homebrew-cask/issues/21913.
android-file-transfer                              rowanj-gitx
javafx-scene-builder (!)                           xquartz


$ brew cask uninstall rowanj-gitx xquartz javafx-scene-builder android-file-transfer
Error: No available formula with the name "brew-cask"
Warning: The default Caskroom location has moved to /usr/local/Caskroom.

Please migrate your Casks to the new location and delete /opt/homebrew-cask/Caskroom,
or if you would like to keep your Caskroom at /opt/homebrew-cask/Caskroom, add the
following to your HOMEBREW_CASK_OPTS:

  --caskroom=/opt/homebrew-cask/Caskroom

For more details on each of those options, see https://github.com/caskroom/homebrew-cask/issues/21913.
Warning: Calling "cask :v1 => 'token'" is deprecated!
Use "cask 'token'" instead.
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cask_loader.rb:36:in `cask'

Warning: Calling MacOS.release is deprecated!
Use MacOS.version instead.
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cask_loader.rb:20:in `block in load'

Warning: Calling MacOS.release is deprecated!
Use MacOS.version instead.
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cask_loader.rb:23:in `block in load'

==> Removing Binary symlink: '/usr/local/bin/gitx'
rowanj-gitx latest is still installed.
Remove it with `brew cask uninstall --force rowanj-gitx`.
Error: xquartz is not installed


$ brew cask uninstall --force rowanj-gitx
Error: No available formula with the name "brew-cask"
Warning: The default Caskroom location has moved to /usr/local/Caskroom.

Please migrate your Casks to the new location and delete /opt/homebrew-cask/Caskroom,
or if you would like to keep your Caskroom at /opt/homebrew-cask/Caskroom, add the
following to your HOMEBREW_CASK_OPTS:

  --caskroom=/opt/homebrew-cask/Caskroom

For more details on each of those options, see https://github.com/caskroom/homebrew-cask/issues/21913.
Warning: Calling "cask :v1 => 'token'" is deprecated!
Use "cask 'token'" instead.
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cask_loader.rb:36:in `cask'


$ find /usr/local/Caskroom -name "*.app" -maxdepth 3
find: /usr/local/Caskroom: No such file or directory


$ brew cask install --force $(brew cask list)

Error: No available formula with the name "brew-cask"
Warning: The default Caskroom location has moved to /usr/local/Caskroom.

Please migrate your Casks to the new location and delete /opt/homebrew-cask/Caskroom,
or if you would like to keep your Caskroom at /opt/homebrew-cask/Caskroom, add the
following to your HOMEBREW_CASK_OPTS:

  --caskroom=/opt/homebrew-cask/Caskroom

For more details on each of those options, see https://github.com/caskroom/homebrew-cask/issues/21913.
Error: No available formula with the name "brew-cask"
Warning: The default Caskroom location has moved to /usr/local/Caskroom.

Please migrate your Casks to the new location and delete /opt/homebrew-cask/Caskroom,
or if you would like to keep your Caskroom at /opt/homebrew-cask/Caskroom, add the
following to your HOMEBREW_CASK_OPTS:

  --caskroom=/opt/homebrew-cask/Caskroom

For more details on each of those options, see https://github.com/caskroom/homebrew-cask/issues/21913.
==> Downloading https://dl.google.com/dl/androidjumper/mtp/current/androidfiletransfer.dmg
######################################################################## 100.0%
==> No checksum defined for Cask android-file-transfer, skipping verification
==> Moving App 'Android File Transfer.app' to '/Applications/Android File Transfer.app'
🍺  android-file-transfer was successfully installed!
Error: undefined method `latest?' for nil:NilClass
Follow the instructions here:
  https://github.com/caskroom/homebrew-cask/blob/master/doc/reporting_bugs/pre_bug_report.md

If this doesn’t fix the problem, please report this bug:
  https://github.com/caskroom/homebrew-cask#reporting-bugs

/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/download.rb:36:in `clear_cache'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/download.rb:14:in `perform'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/installer.rb:96:in `download'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/installer.rb:70:in `install'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/install.rb:25:in `block in install_casks'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/install.rb:19:in `each'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/install.rb:19:in `install_casks'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/install.rb:10:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:110:in `run_command'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:149:in `process'
/usr/local/Homebrew/Library/Homebrew/cask/cmd/brew-cask.rb:8:in `<top (required)>'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
/usr/local/Homebrew/Library/Homebrew/brew.rb:24:in `require?'
/usr/local/Homebrew/Library/Homebrew/brew.rb:100:in `<main>'
```

The only reason I even got into this mess was that roughly every month, I execute `brew doctor` to see if I have to clean up yet another mess due to someone deciding that breaking a perfectly working and stable API seems like a good idea (happened the last time when 'brew upgrade --all' was introduced for no useful reason to the end-user from a stability and workflow reproducibility point of view, only to be removed again a couple of months down the road, and many others in the years I have been using this wonderful system).

Basically, it all started with this harmless:

```
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry and just ignore them. Thanks!

Warning: Your XQuartz (2.7.7) is outdated.
Please install XQuartz 2.7.9 (or delete the current version).
XQuartz can be updated using Homebrew-Cask by running
  brew cask reinstall xquartz
```

This time, it seems to have hit my system harder than other times ;). I am fine to spend some hours trying to figure this mess out myself, if someone would be so kind as to point me into the right direction.

**Comment by moreaki** [comment]
Since my list of casks is very small, I have tried to manually remove them through the brew API. It sort of works:

```
$ brew cask uninstall xquartz   --caskroom=/opt/homebrew-cask/Caskroom
Error: No available formula with the name "brew-cask"
==> Running uninstall process for xquartz; your password may be necessary
==> Removing launchctl service org.macosforge.xquartz.startx
Password:
==> Quitting application ID org.macosforge.xquartz.X11
==> Removing files from pkgutil Bill-of-Materials
==> Removing files: ["/opt/X11/"]
```

Not sure what the weird error message (No available formula with the name "brew-cask") is supposed to tell me; however the directory xquartz is still there, as well as the android-file-transfer one, even after `--force` uninstall those packages for half a dozen times. Other packages are luckily gone:

```
$ find /opt/homebrew-cask/Caskroom -type d -maxdepth 2
/opt/homebrew-cask/Caskroom
/opt/homebrew-cask/Caskroom/android-file-transfer
/opt/homebrew-cask/Caskroom/android-file-transfer/.metadata
/opt/homebrew-cask/Caskroom/android-file-transfer/latest
/opt/homebrew-cask/Caskroom/xquartz
/opt/homebrew-cask/Caskroom/xquartz/2.7.7
```

My question: Can I just remove `/opt/homebrew-cask/Caskroom` and be done with it? I can then cleanly install those packages again afterwards in hope that brew finally decides to work with the new cask directory.

**Comment by vitorgalvao** [comment]
> Can I just remove `/opt/homebrew-cask/Caskroom` and be done with it?

Yes.

Your installation of HBC is broken/outdated. Many clues to that:

* `Error: No available formula with the name "brew-cask"`
* `Warning: Calling "cask :v1 => 'token'" is deprecated!`
* `Warning: Calling MacOS.release is deprecated!``

From the [issue template](https://github.com/caskroom/homebrew-cask/blob/master/.github/ISSUE_TEMPLATE.md) that shows when opening a new issue:

> Before opening your issue **you must check our instructions** for [reporting bugs](https://github.com/caskroom/homebrew-cask#reporting-bugs) (…) [it] has several links to help diagnose and fix common issues.
> (…)
> **if it is apparent you ignored the instructions for reporting issues, we may close the issue without review**.

That seems very much to be the case. Not only will the problem probably be fixed by those instructions, you haven’t given the requested information. We worked hard  on those guides to keep things running smoothly, so I ask that you go back and follow them. It is OK to open a new issue for that.

I understand sometimes we’re tired and don’t want to sift through a guide to make what we feel is a simple bug report. But when each user ignores the guides to save a few minutes, maintainers lose hours and get fatigued. I ask you take that into consideration.

Closing for now, since you mentioned you might just nuke it and start over.

**Comment by moreaki** [comment]
Thank you very much for your time. I have nuked that directory and proceeded to reinstall at least one cask (android-file-transfer), which worked correctly. For the remaining issues that still pertain now, I will read **again** the "reporting bugs" page and open a new issue if warranted.

Last note: The only thing I have done in the past months is "brew update; brew upgrade". How my system can get into this state of what you call broken/outdated is beyond me.

**Comment by vitorgalvao** [comment]
> How my system can get into this state of what you call broken/outdated is beyond me.

May very well be our fault. Some breaking changes coupled with bad timing in updating on your side that screwed something up and you never got the fix.
